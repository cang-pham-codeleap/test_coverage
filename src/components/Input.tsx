import React, { forwardRef } from "react";

/**
 * Props for the Input component
 * @property {string} id - The ID for the input element (required for accessibility)
 * @property {string} [type='text'] - The type of input (text, email, password, etc.)
 * @property {string} [placeholder] - Placeholder text
 * @property {string} [value] - The value of the input
 * @property {string} [defaultValue] - The default value of the input
 * @property {(e: React.ChangeEvent<HTMLInputElement>) => void} [onChange] - Event handler for input changes
 * @property {boolean} [disabled=false] - Whether the input is disabled
 * @property {boolean} [required=false] - Whether the input is required
 * @property {'default' | 'error'} [status='default'] - The visual status of the input
 * @property {string} [errorMessage] - Error message to display when status is 'error'
 * @property {string} [className] - Additional CSS classes
 * @property {React.InputHTMLAttributes<HTMLInputElement>} [props] - Additional HTML attributes
 */
export interface InputProps
  extends Omit<React.InputHTMLAttributes<HTMLInputElement>, "size"> {
  id: string;
  type?: string;
  placeholder?: string;
  value?: string;
  defaultValue?: string;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  disabled?: boolean;
  required?: boolean;
  status?: "default" | "error";
  errorMessage?: string;
  className?: string;
}

/**
 * Input component for form fields
 *
 * @example
 * ```tsx
 * <Input id="email" type="email" placeholder="Enter your email" />
 * <Input
 *   id="password"
 *   type="password"
 *   status="error"
 *   errorMessage="Password must be at least 8 characters"
 * />
 * <Input id="name" defaultValue="John Doe" disabled />
 * ```
 */
export const Input = forwardRef<HTMLInputElement, InputProps>(
  (
    {
      id,
      type = "text",
      placeholder,
      value,
      defaultValue,
      onChange,
      disabled = false,
      required = false,
      status = "default",
      errorMessage,
      className = "",
      ...props
    },
    ref
  ) => {
    // Determine status-specific styles
    const statusClasses = {
      default: "border-gray-300 focus:border-blue-500 focus:ring-blue-500",
      error: "border-red-500 focus:border-red-500 focus:ring-red-500",
    };

    // Create CSS classes based on props
    const inputClasses = [
      "block w-full rounded-md border px-3 py-2 text-sm",
      "shadow-sm placeholder-gray-400",
      "focus:outline-none focus:ring-2",
      disabled ? "bg-gray-100 text-gray-500 cursor-not-allowed" : "bg-white",
      statusClasses[status],
      className,
    ]
      .filter(Boolean)
      .join(" ");

    return (
      <div className="w-full">
        <input
          id={id}
          ref={ref}
          type={type}
          placeholder={placeholder}
          value={value}
          defaultValue={defaultValue}
          onChange={onChange}
          disabled={disabled}
          required={required}
          aria-invalid={status === "error"}
          aria-describedby={status === "error" ? `${id}-error` : undefined}
          className={inputClasses}
          {...props}
        />

        {status === "error" && errorMessage && (
          <p
            id={`${id}-error`}
            className="mt-1 text-sm text-red-500"
            role="alert"
          >
            {errorMessage}
          </p>
        )}
      </div>
    );
  }
);

// Display name for React DevTools
Input.displayName = "Input";

export default Input;
