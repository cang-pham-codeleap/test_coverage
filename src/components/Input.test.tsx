/**
 * Input component tests
 * - Tests basic rendering
 * - Tests different states (default, error, disabled)
 * - Tests event handling
 * - Tests accessibility attributes
 */

import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Input from "./Input";

describe("Input Component", () => {
  // Test basic rendering
  test("renders input element with correct attributes", () => {
    render(<Input id="test-input" />);
    const inputElement = screen.getByRole("textbox");
    expect(inputElement).toHaveAttribute("id", "test-input");
    expect(inputElement).toMatchSnapshot();
  });

  // Test different input types
  test("renders with specified type", () => {
    render(<Input id="email-input" type="email" />);
    const inputElement = screen.getByRole("textbox");
    expect(inputElement).toHaveAttribute("type", "email");
  });

  // Test placeholder
  test("displays placeholder text when provided", () => {
    const placeholder = "Enter your name";
    render(<Input id="name-input" placeholder={placeholder} />);
    const inputElement = screen.getByPlaceholderText(placeholder);
    expect(inputElement).toBeInTheDocument();
  });

  // Test default value
  test("displays default value when provided", () => {
    const defaultValue = "John Doe";
    render(<Input id="name-input" defaultValue={defaultValue} />);
    const inputElement = screen.getByRole("textbox");
    expect(inputElement).toHaveValue(defaultValue);
  });

  // Test value
  test("displays controlled value when provided", () => {
    const value = "Jane Doe";
    render(<Input id="name-input" value={value} onChange={jest.fn()} />);
    const inputElement = screen.getByRole("textbox");
    expect(inputElement).toHaveValue(value);
  });

  // Test disabled state
  test("applies disabled attribute and styling when disabled", () => {
    render(<Input id="disabled-input" disabled />);
    const inputElement = screen.getByRole("textbox");
    expect(inputElement).toBeDisabled();
    expect(inputElement.className).toContain("bg-gray-100");
    expect(inputElement.className).toContain("cursor-not-allowed");
  });

  // Test required attribute
  test("applies required attribute when required", () => {
    render(<Input id="required-input" required />);
    const inputElement = screen.getByRole("textbox");
    expect(inputElement).toBeRequired();
  });

  // Test error status
  test("applies error status styling", () => {
    render(<Input id="error-input" status="error" />);
    const inputElement = screen.getByRole("textbox");
    expect(inputElement.className).toContain("border-red-500");
  });

  // Test error message
  test("displays error message when status is error and message is provided", () => {
    const errorMessage = "This field is required";
    render(
      <Input id="error-input" status="error" errorMessage={errorMessage} />
    );
    const messageElement = screen.getByText(errorMessage);
    expect(messageElement).toBeInTheDocument();
    expect(messageElement).toHaveAttribute("role", "alert");
    expect(messageElement.className).toContain("text-red-500");
  });

  // Test no error message when status is not error
  test("does not display error message when status is default", () => {
    const errorMessage = "This field is required";
    render(
      <Input id="input" status="default" errorMessage={errorMessage} />
    );
    expect(screen.queryByText(errorMessage)).not.toBeInTheDocument();
  });

  // Test onChange handler
  test("calls onChange handler when input value changes", () => {
    const handleChange = jest.fn();
    render(<Input id="test-input" onChange={handleChange} />);
    const inputElement = screen.getByRole("textbox");

    fireEvent.change(inputElement, { target: { value: "new value" } });
    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  // Test accessibility attributes
  test("applies correct aria attributes for error state", () => {
    render(
      <Input 
        id="accessibility-input" 
        status="error" 
        errorMessage="Invalid input" 
      />
    );
    const inputElement = screen.getByRole("textbox");
    expect(inputElement).toHaveAttribute("aria-invalid", "true");
    expect(inputElement).toHaveAttribute("aria-describedby", "accessibility-input-error");
  });

  // Test forwarding ref
  test("forwards ref to the input element", () => {
    const ref = React.createRef<HTMLInputElement>();
    render(<Input id="ref-input" ref={ref} />);
    expect(ref.current).toBeInstanceOf(HTMLInputElement);
    expect(ref.current?.id).toBe("ref-input");
  });

  // Test custom class name
  test("applies custom className when provided", () => {
    const customClass = "custom-input-class";
    render(<Input id="styled-input" className={customClass} />);
    const inputElement = screen.getByRole("textbox");
    expect(inputElement.className).toContain(customClass);
  });
});
