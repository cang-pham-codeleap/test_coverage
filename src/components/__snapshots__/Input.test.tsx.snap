// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component renders input element with correct attributes 1`] = `
<input
  aria-invalid="false"
  class="block w-full rounded-md border px-3 py-2 text-sm shadow-sm placeholder-gray-400 focus:outline-none focus:ring-2 bg-white border-gray-300 focus:border-blue-500 focus:ring-blue-500"
  id="test-input"
  type="text"
  value=""
/>
`;
